submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,197 @@
+## Globalnet 2.0: Enhancements to Globalnet Implementation
+
+## Summary
+
+Currently, Submariner Globalnet implementation in [release 0.8](https://github.com/submariner-io/releases/releases/tag/v0.8.0)
+annotates every Pod, Service and Node in the cluster with a globalIp to support Pod to remote
+Service connectivity in the following manner.
+
+Once a Pod is scheduled and enters Running state, it is annotated with a globalIp. To support
+the connectivity, Globalnet programs the necessary egress rules on the active Gateway Node of
+the cluster where the source IP of the Pod is modified to the corresponding globalIp. This
+translation enables us to uniquely identify the traffic from the Pod, when it reaches the
+destination cluster.
+
+Similarly, a ClusterIP Service is annotated with a globalIp so that remote Pods can communicate
+with it using its globalIp. Globalnet programs the necessary ingress rules on the active Gateway
+node of the cluster, where the service resides. The ingress rules enable any incoming traffic
+destined to the globalIp of the Service to be redirected to the corresponding ClusterIP Service.
+
+In order to support connectivity from the Host Network to the remote services, Globalnet also
+annotates every node in the cluster with a unique globalIp.
+
+While this works fine for small clusters, where the number of Pods and Services are limited, for
+large clusters it poses scalability and performance issues as we would eventually run out of
+globalIps because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
+connectivity as well as Headless Services are not supported.
+
+## Proposal
+
+This enhancement proposes the following changes to make Globalnet implementation scale better.
+
+1. By default, every cluster will be assigned a single globalIp which will be used as egress IP
```

Comment:
This seems like we could easily reach a scale limit due to limited number of ports available.
I suggest we allow either IP or CIDR here to allow good scalability for bigger deployments.

Reply:
That could be achieved by adding per-namespace IPs or selector based IPs, but it's a good point, may be we can reserve a CIDR for that and split the IPs based on the internal cluster CIDRs?

Reply:
That's a good point @mkolesnik. For a single SNAT EgressIP, depending on the configured ephemeral ports, the max number of connections that can be supported will be 65K. But, as you know, this is the active connections at any point of time. Globalnet (or for that matter Submariner) does not maintain the connection info and its handled by Linux Conntrack, so these rules have to be programmed ahead of time. 

IPTables supports providing a contiguous range (not CIDR) while configuring an SNAT rule, so we can specify couple of IPs as globalIPs at ClusterLevel/NamespaceLevel. Have to check how to model this...

Example: iptables -t nat -I SUBMARINER-GN-EGRESS -p all -s 10.0.128.11 -j SNAT --to 169.254.2.11-169.254.2.14

Reply:
For the initial implementation, I feel we can allocate two globalIPs at the cluster level. In future we can make it configurable entity. Your thoughts?